### PR TITLE
Add a page, that lists products on sale

### LINKED ISSUE #15384: [IMPROVEMENT] Add onsale-products listing page
**Scenario**
I think that most of merchants use "ON SALE" flag:
- to get rid of products, that they will no longer sell
- the supplier stopped manufacturing it
- to get rid of slightly damaged item or an item with a missing packaging/manual etc.
- items that have been returned in no-reason period (by law, Czech Republic) and thus cannot be sold normally

People are used to this and when they want to get quality product, usually undamaged, but at lower price, they look for "ON SALE" section of the web page.

**Solution**
Make a page, that lists all products, that have onsale flag turned on.

**Technically**
Basically it's just a copy of newproducts controller. Other than naming, only code change is in Product.php, you are not getting products with `date < 30 days`, but products with `onsale = 1`.
```
add controllers/front/listing/OnsaleProductsController.php
add src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php
add getOnsaleProducts function to classes/Product.php
add a template
add a redirect rule
add a wording
add it to class autoload
```

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR adds a new page that lists all products with "On sale" flag, thus creating a section that can be called "Outlet". It is widely used in European markets to specify goods that are out of production OR you bought so many, that you need to get rid of it. This does not necessarily mean it has some discount on it.
| Type?         | improvement / new feature
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes https://github.com/PrestaShop/PrestaShop/issues/15384
| How to test?  | Tick "On sale" flag on some products and go check out the new "Products on sale" page.

It's my first PR that adds new functionality, but I hope it could be fine!

**Deeper context:**
I think that most of merchants use "ON SALE" (Outlet) flag:
- to get rid of products, that they will no longer sell
- the supplier stopped manufacturing it
- products that are from last-season collection
- to get rid of slightly damaged item or an item with a missing packaging/manual etc.
- items that have been returned and thus cannot be sold normally

People are used to this and when they want to get quality product, usually undamaged, but at lower price, they look for "ON SALE" or "OUTLET" section of the web page. "PRICES DROP" page cannot be used for this, because reducing a price does not mean that the product is going out of sale.

**Screenshots of well known stores that use it:**
SALE

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/19992)

### CODE REVIEW

**Comment by Progi1984** [comment]
Need @PrestaShop/prestashop-product-team validation before continuing on the PR approval ;)

**Comment by Hlavtox** [comment]
@Progi1984 Boss, I updated the PR description 😎

**Comment by marionf** [comment]
Hello @Hlavtox 

Sorry for the delay, there was a misunderstanding about the wording and the goal of this page
Anyway, it seems a good idea and this let the choice to the merchants to display or not this page on the FO as it appears in SEO & URLs in the BO and can be deleted if needed :)

Could you please resolved the conflicts ?
After we will be able to review the wording and the code :)

**Comment by Hlavtox** [comment]
Hi @marionf, I squashed and rebased the PR. :-)

**Comment by marionf** [comment]
Thank you @Hlavtox 
We are going to review the code and the wording now :+1:

File: classes/Product.php
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
    public static function getOnsaleProducts(int $id_lang, int $page_number = 0, int $nb_products = 10, bool $count = false, string $order_by = null, string $order_way = null, Context $context = null)
```

File: classes/Product.php
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        $front = in_array($context->controller->controller_type, ['front', 'modulefront']);
```

File: classes/Product.php
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
            return (int) Db::getInstance((bool) _PS_USE_SQL_SLAVE_)->getValue($sql);
```

File: classes/Product.php
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        $result = Db::getInstance((bool) _PS_USE_SQL_SLAVE_)->executeS($sql);
```

File: classes/Smarty/TemplateFinder.php
Comment (Progi1984) [inline classes/Smarty/TemplateFinder.php:41]:
```suggestion
    /**
     * @var array<int, string>
     */
    private $productListSearchEntities = ['search', 'price-drop', 'best-sale', 'prices-drop', 'best-sales', 'new-products', 'onsale-products'];
```

File: controllers/front/listing/OnsaleProductsController.php
Comment (Progi1984) [inline controllers/front/listing/OnsaleProductsController.php]:
```suggestion
 */

declare(strict_types=1);

use PrestaShop\PrestaShop\Adapter\OnsaleProducts\OnsaleProductsProductSearchProvider;
```

File: controllers/front/listing/OnsaleProductsController.php
Comment (Progi1984) [inline controllers/front/listing/OnsaleProductsController.php]:
```suggestion
    protected function getProductSearchQuery(): ProductSearchQuery
```

File: controllers/front/listing/OnsaleProductsController.php
Comment (Progi1984) [inline controllers/front/listing/OnsaleProductsController.php]:
```suggestion
    protected function getDefaultProductSearchProvider(): OnsaleProductsProductSearchProvider
```

File: controllers/front/listing/OnsaleProductsController.php
Comment (Progi1984) [inline controllers/front/listing/OnsaleProductsController.php]:
```suggestion
    public function getListingLabel(): string
```

File: controllers/front/listing/OnsaleProductsController.php
Comment (Progi1984) [inline controllers/front/listing/OnsaleProductsController.php]:
```suggestion
    public function getBreadcrumbLinks(): array
```

File: src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php
Comment (Progi1984) [inline src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Adapter\OnsaleProducts;
```

File: src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php
Comment (Progi1984) [inline src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php]:
```suggestion
    ): array {
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php
Comment (Progi1984) [inline src/Adapter/OnsaleProducts/OnsaleProductsProductSearchProvider.php]:
```suggestion
    ): ProductSearchResult {
```

**Review by Progi1984 — COMMENTED** [review]
A last feedback

**Review by Progi1984 — APPROVED** [review]

**Comment by Julievrz** [comment]
Hi @Hlavtox,

I discussed this issue with @marionf  and I think there might be an issue with the wording on this page, especially when it comes to translating it. As I speak French, my example will be based on the French translation, but it might concerns other languages as well. 

In French, the 'On sale' flag is translated by 'Promotion'. So if I reduce the price of a product, I'll add the 'On sale' flag to let my customers know that this product has a reduced price. I want to say to my customers: 'Hey the price is reduced, you should have a look' => But this does not mean that I will not sell this product again after it's out. 

But I understand that in other countries the 'On sale' flag means: 'Hurry up, this is your last chance to buy this product because once it's out, it will be gone forever' => This does not mean that the product has a price reduction!

So, I suggest renaming your page 'Last chance' to avoid any misunderstandings. What do you think? I think this suggestion matches the description you have suggested and the purpose of the page.

The last problem is the 'On sale' flag. I think we should differentiate the flag that means ' This product is in promotion' and the one that says 'This is your last chance to buy this product'. At the moment, there is only one. My suggestion is to have two:
'Price drop' flag = the product has a reduction/is in promotion
'Last chance' flag = the product will be gone forever after it's out, it's your last chance to buy it. 

If we keep only 1 flag, the merchant won't be able to use the same flag for two different purposes and it could be confusing, at least in French since 'On sale' means 'Promotion': that would be weird to have a promotion flag if on a page with products that have no reduction => and that's illegal in France anyway.

Marion has opened an issue for this: https://github.com/PrestaShop/PrestaShop/issues/24107

Let us know what you think! 😃

**Comment by Hlavtox** [comment]
Hi @marionf @Julievrz!

If you want to, I think we should just rename "On sale" to "Last chance", together with my new page, and call it done. There is no need to have a second flag.

If the product is discounted, it's said by the discount amount/percentage label and a crossed out price. Why would you add "price drop" label? :-)

**Comment by marionf** [comment]
Hello @Hlavtox 

We have identified the following use case with @Julievrz:

Let's say you have some products discounted and some products at the end of stock but without discount.
You might want to put them forward by displaying the "price drop !" flag on the discounted products and the "last chance !" flag on the end of stock products.

This will be possible if we add a new checkbox on the add/edit product page in BO:
"Display the "Last chance !" flag on the product page, and on product listings." => when this box is checked, then the product is displayed on the new page you created in this PR + the label "Last chance !" is displayed on the product page, and on product listings.

And this way, we keep the "Price drop !" flag for merchants who would like to put forward their products with reduced prices.

**Comment by marionf** [comment]
Hi @Hlavtox 

I described the specs and the behavior for this new "last-chance" page in [this issue](https://github.com/PrestaShop/PrestaShop/issues/24107#issue-861103216)

Let me know if you want to integrate them in your PR ? Otherwise I will close it as the current behavior feet your specific need but isn't the best solution for everyone

**Comment by matks** [comment]
Hello @Hlavtox what do we do with this PR now?

**Comment by PierreRambaud** [comment]
Hi @Hlavtox ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant. If you need help for whatever, don't hesitate to ask ;) 

Thanks!

**Comment by valbewe** [comment]
Hi @Hlavtox. I like the idea and merged your PR.  Unfortunately ps_facetedsearch doesn't work on the new "on-sale" site. 
I've already tried to understand your pull request #https://github.com/PrestaShop/ps_facetedsearch/pull/757 but unfortunately I don't get it to work. 

I extended the function **initializeSupportedControllers()** in **ps_facetedsearch.php** with:

```
'onsale-products' => [
'name' => $this->trans('Onsale products', [], 'Modules.Facetedsearch.Admin'),
'cacheable' => false,
],
```
and **addControllerSpecificFilters()** in **Search.php** with:

```
 if ($this->query->getQueryType() == 'onsale-products') {
     $this->getSearchAdapter()->addFilter('onsale', [0], '>');
}
```
I think I'm failing at MySQL.php. The code should look something like this (I guess): #https://github.com/PrestaShop/ps_facetedsearch/pull/757/files#diff-ca13d1378e6d19627dacdc73c932ac235411edcb94dfb6b93e30662c7181499a 

```
'onsale' => [
                'tableName' => 'product',
                'fieldName' => 'on_sale',
                'joinCondition' => ...
                'joinType' => self::INNER_JOIN,
            ],
```

Are there any other places to edit and can you please help me with this?
In any case:  Thanks and Greetings!

**Pull request closed without merging**
